crates/ruff_python_parser/src/token.rs

```diff
@@ -39,26 +39,24 @@ pub enum Tok {
     String {
         /// The string value.
         value: Box<str>,
-        /// The kind of string.
-        kind: StringKind,
-        /// Whether the string is triple quoted.
-        triple_quoted: bool,
+        /// Flags that can be queried to determine the quote style
+        /// and prefixes of the string
+        flags: StringFlags,
     },
     /// Token value for the start of an f-string. This includes the `f`/`F`/`fr` prefix
     /// and the opening quote(s).
-    FStringStart,
+    FStringStart(StringFlags),
     /// Token value that includes the portion of text inside the f-string that's not
     /// part of the expression part and isn't an opening or closing brace.
     FStringMiddle {
         /// The string value.
         value: Box<str>,
-        /// Whether the string is raw or not.
-        is_raw: bool,
-        /// Whether the string is triple quoted.
-        triple_quoted: bool,
+        /// Flags that can be queried to determine the quote style
+        /// and prefixes of the string
+        flags: StringFlags,
     },
     /// Token value for the end of an f-string. This includes the closing quote.
-    FStringEnd,
+    FStringEnd(StringFlags),
```

Comment:
The information would be required at least for `FStringMiddle` as that contains the string content and is being used by multiple rules. As to `FStringStart` and `FStringEnd`, I think we can get away with not including the information in at least one of them, possibly both. I would check the usages of `FStringStart` / `FStringEnd` in the codebase and see if there are any and check if we can use `FStringMiddle` instead.

Reply:
Hmm, `FStringMiddle` tracking it makes sense when `FStringStart` didn't track it. But it seems redundant now that `FStringStart` exposes the quote information because the prefix (kind) always matches the prefix of the closest `FStringStart` preceding the `FStringMiddle` (that wasn't closed with `FStringEnd`)

Reply:
The issue with f-strings is with empty f-strings. Given an f-string `f''`, here's what the tokenizer currently emits:

```
[
    Ok(
        (
            FStringStart,
            0..2,
        ),
    ),
    Ok(
        (
            FStringEnd,
            2..3,
        ),
    ),
]
```

https://play.ruff.rs/dfe9a07c-8082-4ac8-a199-acf2ea354c92

If we only track quoting style in `FStringMiddle` tokens, that means we won't be able to distinguish between single/double quotes for empty f-strings. Maybe that's an edge case, but the aim of this PR is to make sure that we can always easily retrieve quoting-style information from the tokenizer, and to prepare the ground so that in a future PR we can easily retrieve that information from the AST as well.

I can see a few ways of solving this:
1. Track quoting style in all three f-string tokens: `FStringStart`, `FStringMiddle` and `FStringEnd`. (What this PR currently does.)
2. Only track quoting style in `FStringStart` tokens. That feels like it makes some sense, given that all f-strings currently have a start, but not all of them currently have a middle (in our representation). It would be quite an invasive change relative to what I'm currently doing, however.
3. Start emitting `FStringMiddle` tokens for empty f-strings

Reply:
> (1) seems like the simplest to have for now but I think we can at least not have the information in `FStringEnd` which, I think, is never required.

Good point, I think there are indeed no use cases for carrying the information in `FStringEnd`